worker/src/analytics_engine.rs

```diff
@@ -0,0 +1,186 @@
+use crate::EnvBinding;
+use crate::Result;
+use js_sys::Array;
+use js_sys::Object;
+use std::marker::PhantomData;
+use wasm_bindgen::{JsCast, JsValue};
+use worker_sys::AnalyticsEngineDataset as AnalyticsEngineSys;
+
+#[derive(Clone)]
+pub struct AnalyticsEngineDataset(AnalyticsEngineSys);
+
+unsafe impl Send for AnalyticsEngineDataset {}
+unsafe impl Sync for AnalyticsEngineDataset {}
+
+impl EnvBinding for AnalyticsEngineDataset {
+    const TYPE_NAME: &'static str = "AnalyticsEngineDataset";
+
+    // Override get to perform an unchecked cast from an object.
+    // Miniflare defines the binding as an Object and not a class so its name is not available for checking.
+    // https://github.com/cloudflare/workers-sdk/blob/main/packages/wrangler/templates/middleware/middleware-mock-analytics-engine.ts#L6
+    fn get(val: JsValue) -> Result<Self> {
+        let obj = Object::from(val);
+        Ok(obj.unchecked_into())
+    }
+}
+
+impl JsCast for AnalyticsEngineDataset {
+    fn instanceof(val: &JsValue) -> bool {
+        val.is_instance_of::<AnalyticsEngineDataset>()
+    }
+
+    fn unchecked_from_js(val: JsValue) -> Self {
+        Self(val.into())
+    }
+
+    fn unchecked_from_js_ref(val: &JsValue) -> &Self {
+        unsafe { &*(val as *const JsValue as *const Self) }
+    }
+}
+
+impl From<AnalyticsEngineDataset> for JsValue {
+    fn from(analytics_engine: AnalyticsEngineDataset) -> Self {
+        JsValue::from(analytics_engine.0)
+    }
+}
+
+impl AsRef<JsValue> for AnalyticsEngineDataset {
+    fn as_ref(&self) -> &JsValue {
+        &self.0
+    }
+}
+
+impl AnalyticsEngineDataset {
+    pub fn write_data_point<T>(&self, event: &AnalyticsEngineDataPoint<T>) -> Result<()>
+    where
+        T: BlobType,
+        AnalyticsEngineDataPoint<T>: Clone,
+        JsValue: From<AnalyticsEngineDataPoint<T>>,
+    {
+        Ok(self.0.write_data_point(event.to_js_object()?)?)
+    }
+}
+
+// Marker traits to constrain T
+pub trait BlobType {}
+impl BlobType for String {}
+impl BlobType for Vec<u8> {}
+
+#[derive(Clone, Debug)]
+pub struct AnalyticsEngineDataPoint<T>
+where
+    T: BlobType,
+{
+    indexes: Vec<String>,
+    doubles: Vec<f64>,
+    blobs: Vec<T>,
+}
+
+pub struct AnalyticsEngineDataPointBuilder<T: BlobType> {
+    indexes: Vec<String>,
+    doubles: Vec<f64>,
+    blobs: Vec<T>,
```

Comment:
Notably the values in blobs can be an `Array<bytes | string>` not `Array<bytes> | Array<string>` so our builder methods can mix and match.